src/Connections.Abstractions/ConnectionItems.cs

```diff
@@ -3,14 +3,15 @@
 
 using System;
 using System.Collections;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 
-namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.Internal
+namespace Microsoft.AspNetCore.Connections
 {
-    internal class ConnectionItems : IDictionary<object, object>
+    public class ConnectionItems : IDictionary<object, object>
     {
         public ConnectionItems()
-            : this(new Dictionary<object, object>())
+            : this(new ConcurrentDictionary<object, object>())
```

Comment:
Application models will often be dispatching concurrent messages based on the same connection (SignalR does this today). Unlike `HttpContext`, where the context is a single "thread" of execution for a request, the `ConnectionContext` needs to be accessed by multiple concurrent threads.

I do agree that hiding this behind `IDictionary` kinda defeats the purpose though.

Reply:
Good point, we’ll need to figure out what to do here. The one in SignalR has to be thread safe

Reply:
@anurse That seems to be SignalR specific issue. If SignalR wants to provide a thread-safe abstraction on top of this, it's free to do so. I don't buy the argument that application models built on top of connection handlers are more prone to concurrent access than http middleware.

The HttpConnectionMiddleware is a perfect example of a connection handler that doesn't need or expose an application model that allows concurrent modification of the underlying feature collection, and I can think of many others.

HTTP/2 is another example of why we don't need or want the burden of thread-safety at this layer. Even with the need for stream muxing, it's much better for the connection handler to synchronize access to the underlying connection instead of changing this Dictionary to a ConcurrentDictionary and praying that's sufficient to fix any race conditions.

Reply:
Yeah I’m not arguing for it. My care factor is negative here